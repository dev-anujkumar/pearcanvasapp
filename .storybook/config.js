import { configure } from '@storybook/react'
import '@storybook/addon-console' //Global to print all console errors on storybook canvas  
import { addDecorator } from '@storybook/react'
import { withInfo } from '@storybook/addon-info' //Globle import for info of a story
import { addReadme } from 'storybook-readme'

addDecorator(withInfo)
addDecorator(addReadme)

// automatically import all files ending in *.stories.js under stories folder //
const req = require.context('../stories', true, /\.stories\.js$/)

function loadStories() {
  req.keys().forEach(filename => req(filename))
}

configure(loadStories, module)
