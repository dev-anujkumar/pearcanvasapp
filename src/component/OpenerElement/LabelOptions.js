export const labelOptions = ["No Label", "Chapter", "Ch", "Part", "Unit"]