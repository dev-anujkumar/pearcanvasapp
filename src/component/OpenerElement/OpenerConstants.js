export const labelOptions = ["No Label", "Chapter", "Ch", "Part", "Unit"]

/**
 * Extracts and returns title, number and label from titleText to populate opener element.
 * @param {*} textSemantics Info of position of title, number and label
 * @param {*} type type of data to be extracted
 * @param {*} titleText raw data
 */
export const getOpenerContent = (textSemantics, type, titleText) => {
    let dataToReturn = ""
    if(textSemantics && textSemantics.length > 0 && titleText) {
        let contentData = textSemantics.filter(data => data.type === type)[0]
        let contentNumberData = textSemantics.filter(data => data.type === "number")[0]
        switch(type){
            case "label":
                if(contentData){
                    dataToReturn = titleText.substring(contentData.charStart, contentData.charEnd)
                }
                else{
                    dataToReturn = "No Label"
                }
                break;

            case "number":
                if(contentData){
                    dataToReturn = titleText.substring(contentData.charStart, contentData.charEnd)
                }
                break;

            case "title":
            default:
                if(contentNumberData){
                    dataToReturn = titleText.substring(contentNumberData.charEnd + 1).trimLeft()
                }
                break;      
        }
    }

    return dataToReturn;
}

/**
 * Returns background image data
 * @param {*} path image source url
 */
export const getOpenerImageSource = (path) => {
    if(path && path == "")
        return null
    else
        return path
}
