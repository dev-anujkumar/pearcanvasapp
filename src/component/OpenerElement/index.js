import OpenerElement from './OpenerElement.jsx';
export default OpenerElement;