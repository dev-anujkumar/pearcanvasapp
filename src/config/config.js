let config = {
    REACT_APP_API_URL : process.env.NODE_ENV === "development" ? "https://10.11.7.24:8443/cypress-api/" : '/cypress/canvas-srvr/cypress-api/',
    STRUCTURE_API_URL : process.env.NODE_ENV === "production" ? process.env.STRUCTURE_API_URL : "https://staging.api.pearson.com/",
    LEARNING_OBJECTIVES_ENDPOINT: process.env.NODE_ENV === "production" ? process.env.LEARNING_OBJECTIVES_ENDPOINT : "https://contentapis-staging.pearsoncms.net/lo-api/",
    ASSET_POPOVER_ENDPOINT: process.env.NODE_ENV === "production" ? process.env.ASSET_POPOVER_ENDPOINT : "https://contentapis-staging.pearsoncms.net/manifest-api/",
    PRODUCTAPI_ENDPOINT: process.env.NODE_ENV === "production" ? process.env.PRODUCTAPI_ENDPOINT : "https://contentapis-staging.pearsoncms.net/product-api/",
    ELM_ENDPOINT: process.env.NODE_ENV === "production" ? process.env.ELM_ENDPOINT : "https://contentapis-staging.pearsoncms.net/manifest-api/",
    SLATE_REFRESH_URL : process.env.NODE_ENV === "production" ? process.env.SLATE_REFRESH_URL :"https://contentapis-staging.pearsoncms.net/structure-api/container/v2/",
    AUDIO_NARRATION_URL : process.env.NODE_ENV === "production" ? process.env.STRUCTURE_API_ENDPOINT_URL : 'https://contentapis-staging.pearsoncms.net/structure-api/',
    AUDIO_API_KEY :process.env.NODE_ENV === "production" ? process.env.AUDIO_API_KEY : "Gf7G8OZPaVGtIquQPbqpZc6D2Ri6A5Ld",
    API_URL: "./api",
    NODE_ENV: process.env.NODE_ENV || 'development',
    COREAPI_ENDPOINT: process.env.NODE_ENV === "production" ? process.env.COREAPI_ENDPOINT : "https://contentapis-staging.pearsoncms.net/core-api",
    JAVA_API_URL: process.env.NODE_ENV === 'production' ? process.env.JAVA_API_URL : "https://10.11.7.24:8443/app/toc-javaapp/",
    NARRATIVE_API_URL: "https://10.11.7.24:8443/app/toc-javaapp/v1/",
    STRUCTURE_APIKEY: process.env.NODE_ENV === 'production' ? process.env.STRUCTURE_APIKEY:'Gf7G8OZPaVGtIquQPbqpZc6D2Ri6A5Ld',
    MANIFEST_APIKEY: process.env.NODE_ENV === 'production' ? process.env.MANIFEST_APIKEY: 'YFeLXDGqbBj2GZf85jpcZOQCEasAK5hc',
    ssoToken: "i2KQRvTIpS5MCcQniBc38hsfEgc.*AAJTSQACMDIAAlNLABx6ODZncW9ZcmlwMGlsN2IwUHNlK0lVeEk4TlE9AAJTMQACMDE.*",
    alfrescoMetaData : {},
    slateEntityURN : "urn:pearson:entity:2b03e70f-8730-451b-9f9a-b496b6d91c9e",
    slateManifestURN : "urn:pearson:manifest:d9023151-3417-4482-8175-fc965466220e",
    parentContainerUrn:"",
    parentEntityUrn:"",
    slateType : 'section',
    currentInsertedIndex : 0,
    currentInsertedType : "",
    disableNext : false,
    disablePrev : false,
    //c2 required keys and urls
    CMDS_APIKEY: process.env.NODE_ENV === 'production' ? process.env.CMDS_APIKEY : '5x8gLqCCfkOfgPkFd9YNotcAykeldvVd',
    CMDS_DATA_ENDPOINT: process.env.NODE_ENV === 'production' ? process.env.CMDS_DATA_ENDPOINT : 'https://staging.data.pearson.com',
    CMDS_SCHEMA_ENDPOINT: process.env.NODE_ENV === 'production' ? process.env.CMDS_SCHEMA_ENDPOINT : 'https://staging.schema.pearson.com',
    CMDS_DATABASE: process.env.NODE_ENV === 'production' ? process.env.CMDS_DATABASE : '?db=qa2',
    CMIS_REPO   : process.env.NODE_ENV === 'production' ? process.env.CMIS_REPO : '[{"repo":"https://staging.api.pearson.com/content/cmis/ukwip","repoName":"UK"},{"repo":"https://staging.api.pearson.com/content/cmis/uswip-aws","repoName":"AWS US"},{"repo":"https://staging.api.pearson.com/content/cmis/uswip","repoName":"US"}]',
    EPS_API: process.env.NODE_ENV === 'production' ? process.env.EPS_API : 'https://us-school-stg.pearsoned.com/school',
    //c4 required keys and urls
    CTOOL_APIKEY: process.env.NODE_ENV === 'production' ? process.env.CTOOL_APIKEY : 'INpAP0MaiVP9Rohy4rB0ue3j2SGE3Abk',
    CTOOL_PUBSLATE: process.env.NODE_ENV === 'production' ? process.env.CTOOL_PUBSLATE : 'https://staging.api.pearson.com/content/tools/wip2cite/publishslate/v1/',
    CTOOL_PUBTITLE: process.env.NODE_ENV === 'production' ? process.env.CTOOL_PUBTITLE : 'https://staging.api.pearson.com/content/tools/wip2cite/publishtitle/v1/',
    CTOOL_PUBTITLES3: process.env.NODE_ENV === 'production' ? process.env.CTOOL_PUBTITLES3 : 'https://staging.api.pearson.com/content/tools/wip2cite/publishtitles3/v1/',
    CTOOL_DISCIPLINEID: process.env.NODE_ENV === 'production' ? process.env.CTOOL_DISCIPLINEID : 'https://staging.api.pearson.com/content/tools/printondemand/disciplineid/v1/',
    CTOOL_REGISTERPOD: process.env.NODE_ENV === 'production' ? process.env.CTOOL_REGISTERPOD : 'https://staging.api.pearson.com/content/tools/printondemand/registerpod/v1/',
    CTOOL_DASHBOARD: process.env.NODE_ENV === 'production' ? process.env.CTOOL_DASHBOARD : 'https://staging.api.pearson.com/content/tools/chaucerdashboard/v1', 
    EPUB_ENDPOINT: process.env.NODE_ENV === 'production' ? process.env.EPUB_ENDPOINT : 'https://staging.api.pearson.com/content/tools/transformation/narrative/v1',
    C6PUB_ENDPOINT: process.env.NODE_ENV === 'production' ? process.env.C6PUB_ENDPOINT : 'https://staging.api.pearson.com/content/delivery/publish/v2/',
    C6PUB_API_KEY: process.env.NODE_ENV === 'production' ? process.env.C6PUB_API_KEY : '5x8gLqCCfkOfgPkFd9YNotcAykeldvVd',
    C6REDIS_SERVER_UPDATE: 'api/projects/',
    C4_API_URL: "./api",
    projectUrn: "urn:pearson:distributable:977c95a8-e16a-413c-bfd0-788fd2a3698d",
    projectEntityUrn:"urn:pearson:entity:3d9363f1-36bb-47ea-8842-9b142027692c",
    citeUrn:"urn:pearson:manifestation:7fa4ae52-fabc-4a7f-8876-6054f33d36c4",
    WRAPPER_URL: process.env.NODE_ENV === 'production' ? `${window.parent.origin}/toc-wrapper/index.html` : 'https://localhost:4000',
    book_title:"ELMTEST_StgEnv_Krajewski Test",
    IDENTITY_URL: process.env.NODE_ENV === 'production' ? process.env.IDENTITY_URL : "/auth",
    //parentUrl: window.location.origin
    // https://dev-structuredauthoring.pearson.com/cypress/dashboard-srvr
    LOCK_API_BASE_URL : process.env.NODE_ENV === 'production' ? `${window.parent.origin}/cypress/dashboard-srvr` : 'https://dev-structuredauthoring.pearson.com/cypress/dashboard-srvr',
    PATTERNS: {
        PATTERN_ADD_ASSET: process.env.NODE_ENV === 'production' ? process.env.PATTERN_ADD_ASSET : 'https://component-lib-stg.pearson.com/c2/654b2512-649f-42ab-9c14-72cf4ce380f7/PatternAddAnAsset.js',
        PATTERN_BROKER: process.env.NODE_ENV === 'production' ? process.env.PATTERN_BROKER : 'https://component-lib-stg.pearson.com/c2/7a03593e-61b0-4d72-ab3c-4fdd5d14ad06/PatternBroker.js',
        PATTERN_PRODUCT_LINK: process.env.NODE_ENV === 'production' ? process.env.PATTERN_PRODUCT_LINK : 'https://component-lib-stg.pearson.com/c2/4e6724b9-b65e-41ac-a132-de949cec3948/PatternProductLink.js',
        PATTERN_VENDOR: process.env.NODE_ENV === 'production' ? process.env.PATTERN_VENDOR : 'https://component-lib-stg.pearson.com/c2/6004cda8-7f38-4377-b7a6-5d06184a5de5/Patternvendor.js',
        PATTERN_SEARCH_SELECT: process.env.NODE_ENV === 'production' ? process.env.PATTERN_SEARCH_SELECT : 'https://component-lib-stg.pearson.com/c2/854fdf48-456c-4021-8ffb-2d9c969e50d4/PatternSearchSelect.js',
    },
    colors : ["#000000", "#003057", "#505759", "#005A70", "#006128"],
    isCO : false,
    isLOL:false,
    toolBarList : ['bold','italic','underline','strikethrough','clearformatting','increaseindent','decreaseindent','footnote','glossary','orderedlist','unorderedlist','mathml','chemml','inlinecode','superscript','subscript','specialcharactor','undo','redo','assetpopover','slatetag'],
    elementToolbar: [],
    headingToolbar : ['italic','clearformatting','increaseindent','footnote','mathml','chemml','superscript','subscript','specialcharactor','undo','redo','assetpopover','slatetag'],
    codeListingToolbar: ['bold','italic','underline','strikethrough','clearformatting','decreaseindent','footnote','glossary','orderedlist','unorderedlist','mathml','chemml','inlinecode','superscript','subscript','specialcharactor','undo','redo','assetpopover','slatetag'],
    asideToolbar: ['bold','italic','underline','strikethrough','clearformatting','decreaseindent','footnote','glossary','orderedlist','unorderedlist','mathml','chemml','inlinecode','superscript','subscript','specialcharactor','undo','redo','assetpopover','slatetag'],
    labelToolbar:['footnote','decreaseindent'],
    captionToolbar:['decreaseindent'],
    GET_FIGURES : process.env.NODE_ENV === 'production' ? process.env.GET_FIGURES : 'https://contentapis-qa.pearsoncms.net/',
    GET_ASSETPOPOVER_ID : process.env.NODE_ENV === 'production' ? process.env.GET_ASSETPOPOVER_ID :'https://staging.api.pearson.com/',
    APO_API_KEY : process.env.NODE_ENV === 'production' ? process.env.APO_API_KEY : '7ij8zrLkCNR9DtAXx3KJ6yutoYnk4SAx'
};
if (process.env.NODE_ENV === "development") {
    config.userName = 'c5test01';
    config.userId= 'c5test01';
    config.userEmail = 'c5test01@mctest.local';
    config.assignee='c5test01';
}

export default config;
