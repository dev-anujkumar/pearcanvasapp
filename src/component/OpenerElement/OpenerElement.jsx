import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { labelOptions, getOpenerContent, getOpenerImageSource } from './OpenerConstants'
import { dropdownArrow } from './../../images/ElementButtons/ElementButtons.jsx';

import '../../styles/OpenerElement/OpenerElement.css'
import { hasReviewerRole } from '../../constants/utility';
import noImage from '../../images/OpenerElement/no-image.png'
import { c2MediaModule } from './../../js/c2_media_module';

import { hideTocBlocker, disableHeader } from '../../js/toggleLoader'
import config from '../../config/config';
import { checkSlateLock } from "../../js/slateLockUtility.js"
import axios from 'axios';

class OpenerElement extends Component {

    constructor(props){
        super(props)
        const { textsemantics, text } = props.element.title;
        const bgImage = props.element.backgroundimage.path;

        if (document.querySelector("[name='alt_text']") && props.element.backgroundimage.alttext)
            document.querySelector("[name='alt_text']").innerHTML = props.element.backgroundimage.alttext;
        if (document.querySelector("[name='long_description']") && props.element.backgroundimage.longdescripton)
            document.querySelector("[name='long_description']").innerHTML = props.element.backgroundimage.longdescripton;

        this.state = {
            label: getOpenerContent(textsemantics, "label", text) || "No Label",
            number: getOpenerContent(textsemantics, "number", text),
            title: getOpenerContent(textsemantics, "title", text),
            showLabelDropdown: false,
            imgSrc: getOpenerImageSource(bgImage),
            width: null,
            imageId: props.element.backgroundimage.imageid ? props.element.backgroundimage.imageid : "",
            projectMetadata: false
        }
    }

    dataFromAlfresco = (data) => {
        hideTocBlocker();
        disableHeader(false);
        let imageData = data;
        let epsURL = imageData['EpsUrl'] ? imageData['EpsUrl'] : "";
        let imageId = imageData['workURN'] ? imageData['workURN'] : "";
        let figureType = imageData['assetType'] ? imageData['assetType'] : "";
        let width = imageData['width'] ? imageData['width'] : "";
        let smartLinkString = (imageData.desc && imageData.desc.toLowerCase() !== "eps media") ? imageData.desc : "{}";
        let smartLinkDesc = smartLinkString !== "{}" ? JSON.parse(smartLinkString) : "";
        let smartLinkType = smartLinkDesc !== "" ? smartLinkDesc.smartLinkType : "";
        if (figureType === "image" || figureType === "table" || figureType === "mathImage" || figureType === "authoredtext") {
            let altText = imageData['alt-text'] ? imageData['alt-text'] : "";
            let longDesc = imageData['longDescription'] ? imageData['longDescription'] : "";
            this.setState({
                imgSrc: epsURL,
                imageId: imageId,
                width
            });
            if (document.querySelector("[name='alt_text']"))
                document.querySelector("[name='alt_text']").innerHTML = altText;
            if (document.querySelector("[name='long_description']"))
                document.querySelector("[name='long_description']").innerHTML = longDesc;
        }
        
        this.handleBlur({imgSrc: epsURL, imageId});
        disableHeader(false)
        hideTocBlocker()
    }
    handleC2ExtendedClick = (data) => {
        let data_1 = data;
        let that = this;
        !hasReviewerRole() && c2MediaModule.productLinkOnsaveCallBack(data_1, function (data_2) {
            c2MediaModule.AddanAssetCallBack(data_2, function (data) {
                that.dataFromAlfresco(data);
            })
        })
    }

     /**
     * Responsible for opening C2 media popup
     * @param {e} event
     */
    handleC2MediaClick = (e) => {
        if(hasReviewerRole()){
            return true
        }
        const { slateLockInfo , permissions } = this.props
        if(checkSlateLock(slateLockInfo))
            return false

        if (e.target.tagName.toLowerCase() === "p") {
            e.stopPropagation();
            return;
        }
        let that = this;
        let alfrescoPath = config.alfrescoMetaData;
        if (alfrescoPath && this.state.projectMetadata) {
            alfrescoPath.alfresco = this.state.projectMetadata.alfresco;
        }
        var data_1 = false;
        if(alfrescoPath && alfrescoPath.alfresco && Object.keys(alfrescoPath.alfresco).length > 0 ) {
        if (alfrescoPath.alfresco.nodeRef) {
            if(this.props.permissions && this.props.permissions.includes('add_multimedia_via_alfresco'))    { 
            data_1 = alfrescoPath.alfresco;
            /*
                data according to new project api 
            */
            data_1['repositoryName'] = data_1['repoName'] ? data_1['repoName'] : data_1['repositoryName']
            data_1['repositoryFolder'] = data_1['name'] ? data_1['name'] : data_1['repositoryFolder']
            data_1['repositoryUrl'] = data_1['repoInstance'] ? data_1['repoInstance'] : data_1['repositoryUrl']
            data_1['visibility'] = data_1['siteVisibility'] ? data_1['siteVisibility'] : data_1['visibility']
            /*
                data according to old core api and c2media
            */
            data_1['repoName'] = data_1['repositoryName'] ? data_1['repositoryName'] : data_1['repoName']
            data_1['name'] = data_1['repositoryFolder'] ? data_1['repositoryFolder'] : data_1['name']
            data_1['repoInstance'] = data_1['repositoryUrl'] ? data_1['repositoryUrl'] : data_1['repoInstance']
            data_1['siteVisibility'] = data_1['visibility'] ? data_1['visibility'] : data_1['siteVisibility']
            this.handleC2ExtendedClick(data_1)
            }
            else{
                this.props.accessDenied(true)
            }
        }
        } else {
            if (this.props.permissions.includes('alfresco_crud_access')) {
                c2MediaModule.onLaunchAddAnAsset(function (alfrescoData) {
                    data_1 = { ...alfrescoData };
                    let request = {
                        eTag: alfrescoPath.etag,
                        projectId: alfrescoPath.id,
                        ...alfrescoPath,
                        additionalMetadata: { ...alfrescoData },
                        alfresco: { ...alfrescoData }
                    };

                    /*
                        preparing data according to Project api
                    */

                    request.additionalMetadata['repositoryName'] = data_1['repoName'];
                    request.additionalMetadata['repositoryFolder'] = data_1['name'];
                    request.additionalMetadata['repositoryUrl'] = data_1['repoInstance'];
                    request.additionalMetadata['visibility'] = data_1['siteVisibility'];

                    request.alfresco['repositoryName'] = data_1['repoName'];
                    request.alfresco['repositoryFolder'] = data_1['name'];
                    request.alfresco['repositoryUrl'] = data_1['repoInstance'];
                    request.alfresco['visibility'] = data_1['siteVisibility'];

                    that.handleC2ExtendedClick(data_1)
                    /*
                        API to set alfresco location on dashboard
                    */
                    let url = config.PROJECTAPI_ENDPOINT + '/' + request.projectId + '/alfrescodetails';
                    let SSOToken = request.ssoToken;
                    return axios.patch(url, request.alfresco,
                        {
                            headers: {
                                'Accept': 'application/json',
                                'ApiKey': config.STRUCTURE_APIKEY,
                                'Content-Type': 'application/json',
                                'PearsonSSOSession': SSOToken,
                                'If-Match': request.eTag
                            }
                        })
                        .then(function (response) {
                            let tempData = { alfresco: alfrescoData };
                            that.setState({
                                projectMetadata: tempData
                            })
                        })
                        .catch(function (error) {
                            console.log("error", error)
                        });
                })
            }
            else {
                this.props.accessDenied(true)
            }
        }
    }

    /**
     * Handles label model change event
     * @param {e} event
     */
    handleOpenerLabelChange = e => {
        this.setState({
            label: e.target.innerHTML
        })
        this.handleBlur(e)
        this.toggleLabelDropdown()
    }
    
    /**
     * Toggles label dropdown
     */
    toggleLabelDropdown = (e) => {
        this.setState({
            showLabelDropdown: !this.state.showLabelDropdown
        })
    }

    /**
     * Renders label dropdown
     */
    renderLabelDropdown = () => {
        const { showLabelDropdown } = this.state
        const openerLabelOptions = labelOptions.map((value, index) => {
            return <li key={index} data-value={value} onClick={this.handleOpenerLabelChange}>{value}</li>
        })
        if(showLabelDropdown){
            return (
                <ul className="element-dropdown-content">
                    {openerLabelOptions}
                </ul>
            )
        }
        else{
            return null
        }
    }

    /**
     * Handles number model change event
     * @param {e} event
     */
    handleOpenerNumberChange = e => {
        this.setState({
            number: e.target.value
        })
    }

    /**
     * Handles title model change event
     * @param {e} event
     */
    handleOpenerTitleChange = e => {
        this.setState({
            title: e.target.value            
        })
    }

    /**
     * Validates input in the number field
     * @param {e} event
     */
    numberValidatorHandler = (e) => {
        let charCode = (e.which) ? e.which : e.keyCode;
        if((charCode>=48 && charCode<= 57) || (charCode>=65 && charCode<=90) || (charCode>=97 && charCode<=122)){
            return true
        } else {
            e.preventDefault()
            return false
        }
    }
    /**
     * Handles background image style
     *  @param {imgSrc} image source
     */
    getBGStyle = (imgSrc, width) => {
        let styleObj = {
            "width" : "24%"
        }
        if (imgSrc) {
            //  styleObj["width"] = width ? width : "100%"
              styleObj = {
              "max-width": "605px",
              "height": "384px",
              "margin-top": "1.5em",
              "margin-left": "auto",
             "margin-right": "auto",
             "width":"76%"
          }
        }

        return styleObj
    }
    
    /**
     * Handles Focus on opener element
     * @param {slateLockInfo} Slate lock data
     */
    handleOpenerClick = (slateLockInfo, e) => {
        if(checkSlateLock(slateLockInfo)){
            e.preventDefault()
            return false
        }
        this.props.onClick()

    }

    createSemantics = ({...values}) => {
        let textSemantics = [];
        let currentIndex = 0;
        
        Object.keys(values).forEach(item => {
            textSemantics.push({
                "type": item,
                "charStart": currentIndex,
                "charEnd": currentIndex += (values[item]).length
            });
            currentIndex++;
        });

        return textSemantics;
    }

    /**
     * Handles blur event for each input box and initiates saving call
     * @param {*} event blur event object
     */
    handleBlur = (event) => {
        if(checkSlateLock(this.props.slateLockInfo)){
            event.preventDefault()
            return false
        }

        /**
         * [BG-411]|7 - validate before making blur call
         */
        const { textsemantics, text } = this.props.element.title;
        const classList = event.currentTarget && event.currentTarget.classList || [];
        let flag = true;
        if (classList.length > 0 
            && (classList.contains("opener-title") || classList.contains("opener-number"))
            && (this.state.number === getOpenerContent(textsemantics, "number", text))
            && (this.state.title === getOpenerContent(textsemantics, "title", text))) {
            flag = false;
        }

        let element = this.props.element;
        let { label, number, title, imgSrc, imageId } = this.state;
        label = event.target && event.target.innerText ? event.target.innerText : label;
        imgSrc = event.imgSrc || imgSrc;
        imageId = event.imageId || imageId;

        if(!element.title) {
            element.title = {
                "schema": "http://schemas.pearson.com/wip-authoring/authoredtext/1#/definitions/authoredtext",
                "text": "",
                "textsemantics": []
            };
        }
                
        element.title.text = `${label} ${number}: ${title}`;
        element.title.textsemantics = this.createSemantics({label, number});

        if(!element.backgroundimage) {
            element.backgroundimage = {
                "path": "",
                "schema": "http://schemas.pearson.com/wip-authoring/image/1#/definitions/image",
                "imageid": "",
                "alttext": "",
                "longdescripton": "",
                "credits": {
                    "schema": "http://schemas.pearson.com/wip-authoring/authoredtext/1#/definitions/authoredtext",
                    "text": ""
                }
            };
        }

        let altText = "";
        let longDesc = "";
        if (document.querySelector("[name='alt_text']"))
            altText = document.querySelector("[name='alt_text']").innerHTML;
        if (document.querySelector("[name='long_description']"))
            longDesc = document.querySelector("[name='long_description']").innerHTML;

        element.backgroundimage.path = imgSrc;
        element.backgroundimage.imageid = imageId;
        element.backgroundimage.alttext = altText;
        element.backgroundimage.longdescription = longDesc;
        element.backgroundcolor = this.props.backgroundColor;

        flag && this.props.updateElement(element);
    }
    
    
    render() {
        const { imgSrc, width } = this.state
        const { element, backgroundColor, slateLockInfo } = this.props
        let isDisable = hasReviewerRole() ? " disable-role" : ""
        const styleObj = this.getBGStyle(imgSrc, width)
        return (
            <div className = "opener-element-container" onClickCapture={(e) => this.handleOpenerClick(slateLockInfo, e)}>
                <div className = "input-box-container">
                    <div className="opener-label-box">
                        <div className="opener-label-text">Label</div>
                        <div className={"element-dropdown-title label-content" + isDisable} onClick={this.toggleLabelDropdown}>{this.state.label}
                            {this.renderLabelDropdown()}
                            <span>{dropdownArrow}</span>
                        </div>
                    </div>
                    <div className="opener-label-box oe-number-box">
                        <div className="opener-number-text">Number</div>
                        <input className={"element-dropdown-title opener-number" + isDisable} maxLength="9" value={this.state.number} type="text" onChange={this.handleOpenerNumberChange} onKeyPress={this.numberValidatorHandler} onBlur={this.handleBlur} />
                    </div>
                    <div className="opener-label-box oe-title-box">
                        <div className="opener-title-text">Title</div>
                        <input className={"element-dropdown-title opener-title" + isDisable} value={this.state.title} type="text" onChange={this.handleOpenerTitleChange} onBlur={this.handleBlur} />
                    </div>
                </div>
                <figure className="pearson-component opener-image figureData" onClick={this.handleC2MediaClick} style={{ backgroundColor: `${backgroundColor}` }}>
                    <img style={styleObj} src={imgSrc ? imgSrc : noImage}
                        draggable="false" 
                    />
                </figure>
            </div>
        )
    }
}

OpenerElement.propTypes = {
    /** Model */
    model : PropTypes.object.isRequired,
}

OpenerElement.displayName = 'OpenerElement'

export default OpenerElement;
  