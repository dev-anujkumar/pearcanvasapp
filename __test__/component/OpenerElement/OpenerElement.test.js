import React from 'react';
import { mount } from 'enzyme';
import OpenerElement from '../../../src/component/OpenerElement';
import { openerElementData } from '../../../fixtures/OpenerElementData'
import config from '../../../src/config/config';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
    appStore: {
        permissions: [
            "login", "logout", "bookshelf_access", "generate_epub_output", "demand_on_print", "toggle_tcm", "content_preview", "add_instructor_resource_url", "grid_crud_access", "alfresco_crud_access", "set_favorite_project", "sort_projects",
            "search_projects", "project_edit", "edit_project_title_author", "promote_review", "promote_live", "create_new_version", "project_add_delete_users", "create_custom_user", "toc_add_pages", "toc_delete_entry", "toc_rearrange_entry", "toc_edit_title", "elements_add_remove", "split_slate", "full_project_slate_preview", "access_formatting_bar",
            "authoring_mathml", "slate_traversal", "trackchanges_edit", "trackchanges_approve_reject", "tcm_feedback", "notes_access_manager", "quad_create_edit_ia", "quad_linking_assessment", "add_multimedia_via_alfresco", "toggle_element_page_no", "toggle_element_borders", "global_search", "global_replace", "edit_print_page_no", "notes_adding", "notes_deleting", "notes_delete_others_comment", "note_viewer", "notes_assigning", "notes_resolving_closing", "notes_relpying",
        ]
    },

});

describe('Testing Opener component with props', () => {
    const props = {
        slateLockInfo: {
            isLocked: false,
            userId: 'c5Test01'
        },
        model : openerElementData.html,
        onClick : ()=>{},
        permissions: []
    }
    it('Simulating click event to open label dropdown', () => {
        const openerComponent = mount( <Provider store={store}><OpenerElement {...props} /></Provider> )
        openerComponent.find('div.element-dropdown-title.label-content').simulate('click');
        openerComponent.find('ul.element-dropdown-content>li:first-child').simulate('click');
    })
    it('Changing input number', () => {
        const openerComponent = mount( <Provider store={store}><OpenerElement {...props} /></Provider> )
        openerComponent.find('input.element-dropdown-title.opener-number').simulate('change', { target: { value: '1234567890!!!' } });
    })
    describe('Simulating keyPress event on input number', () => {
        const openerComponent = mount( <Provider store={store}><OpenerElement  /></Provider> )
        it('Simulating keyPress event on input number - alphanumeric input', () => {
            openerComponent.find('input.element-dropdown-title.opener-number').simulate('keypress', { which: 48 });
        })
        it('Simulating keyPress event on input number - special character input', () => {
            openerComponent.find('input.element-dropdown-title.opener-number').simulate('keypress', { which: 91 });
        })
    })
    it('Changing input title', () => {
        const openerComponent = mount( <Provider store={store}><OpenerElement {...props} /></Provider> )
        openerComponent.find('input.element-dropdown-title.opener-title').simulate('change', { target: { value: '1234567890!!!' } });
    })
    it('Simulating alfresco click without alfresco location', () =>{
        const openerComponent = mount( <Provider store={store}><OpenerElement {...props} /></Provider> )
        openerComponent.find('OpenerElement').instance().handleC2MediaClick({target : {tagName : 'b'}}) 
    })
    it('Simulating alfresco click with alfresco location', () =>{
        const openerComponent = mount( <Provider store={store}><OpenerElement {...props} /></Provider> )
        config.alfrescoMetaData = {nodeRef : {}}
        openerComponent.find('OpenerElement').instance().handleC2MediaClick({target : {tagName : 'b'}}) 
    })
    it('Alfresco Data Handling', () => {
        const openerComponent = mount(<Provider store={store}><OpenerElement {...props} /></Provider>, { attachTo: document.body })
        openerComponent.find('OpenerElement').instance().dataFromAlfresco({ assetType: "image" })
    })
})
