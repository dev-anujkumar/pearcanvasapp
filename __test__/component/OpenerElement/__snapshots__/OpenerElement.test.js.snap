// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Opener component with props render opener element component  1`] = `
<OpenerElement
  model={
    Object {
      "label": "Ch",
      "number": "166",
      "title": "Opener element title 1",
    }
  }
>
  <div
    className="opener-element-container"
  >
    <div
      className="input-box-container"
    >
      <div
        className="opener-label-box"
      >
        <div
          className="opener-label-text"
        >
          Label
        </div>
        <div
          className="element-dropdown-title label-content"
          onClick={[Function]}
        >
          Ch
          <span>
            <svg
              className="dropdown-arrow"
              viewBox="0 0 9 4.5"
            >
              <path
                d="M0,0,4.5,4.5,9,0Z"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        className="opener-label-box"
      >
        <div
          className="opener-number-text"
        >
          Number
        </div>
        <input
          className="element-dropdown-title opener-number"
          maxLength="9"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value="166"
        />
      </div>
      <div
        className="opener-label-box"
      >
        <div
          className="opener-title-text"
        >
          Title
        </div>
        <input
          className="element-dropdown-title opener-title"
          onChange={[Function]}
          type="text"
          value="Opener element title 1"
        />
      </div>
    </div>
    <figure
      className="pearson-component opener-image figureData"
    >
      <img
        draggable="false"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </figure>
  </div>
</OpenerElement>
`;
